### ISSUE TITLE
[clang++-18] Unexpected `deprecated-declarations` warning, when this warning is explicitly suppressed

### ISSUE THREAD

**Issue by ralfkonrad** [llvm/llvm-project#110321]
`clang-18` shows an unexpected `deprecated-declarations` warning when a deprecated constructor is used in `std::make_shared<T>(args from deprecated constructor...)` while the warning is turned of using `_Pragma("clang diagnostic ignored \"-Wdeprecated-declarations\"")`.

On the other hand using `std::shared_ptr<T>(new T(args from deprecated constructor...))` does not emit any warning as expected when it off.

Neiter `clang-16` and `clang-17` nor `gcc` on `ubuntu-24.04` and `msvc` emit a warning.

The same behaviour can be seen with `boost::shared_ptr` instead of `std::shahred_ptr`.

To reproduce this issue, see https://github.com/ralfkonrad/clang_deprecation_warning and its [workflow results](https://github.com/ralfkonrad/clang_deprecation_warning/actions).

**Comment by shafik** [comment]
Seems we don't have access to the repo but a gotbolt link w/ a minimal reproducer is almost always a better option.

**Comment by ralfkonrad** [comment]
I have reproduced the warning in a small example: https://godbolt.org/z/vd5vP6c4P

I do not understand the warning `warning: 'Data' is deprecated [-Wdeprecated-declarations]` emitted from line 31 as it is guarded by 
```cpp
_Pragma("clang diagnostic push")
    _Pragma("clang diagnostic ignored \"-Wdeprecated-declarations\"")
        Class ::Class(int i)
    : data_(std::make_shared<Data>(i)) {}
_Pragma("clang diagnostic pop")
```

However, line 73
```cpp
_Pragma("clang diagnostic push")
    _Pragma("clang diagnostic ignored \"-Wdeprecated-declarations\"")
        Class ::Class(int i)
    : data_(std::shared_ptr<Data>(new Data(i))) {}
_Pragma("clang diagnostic pop")
```
does not emit a warning.

I would expect that `std::make_shared<Data>(i)` and the "traditional" `std::shared_ptr<Data>(new Data(i))` behave the same way here.

This behavior is there in `clang++-18` and `clang++-19`, but older version do not emit a deprecation warning.
